rust/cardano-blockchain-types/src/cip36/key_registration.rs

```diff
@@ -0,0 +1,360 @@
+            let pub_key = decode_bytes(d, "CIP36 Key Registration voting key, single voting key")?;
+            let vk = voting_pk_vec_to_verifying_key(&pub_key).map_err(|e| {
+                decode::Error::message(format!(
+                    "CIP36 Key Registration voting key, singe voting key, {e}"
+                ))
+            })?;
+            // Since there is 1 voting key, all the weight goes to this key = 1.
+            voting_keys.push(VotingPubKey {
+                voting_pk: vk,
+                weight: 1,
+            });
+        },
+        // CIP36 type registration (multiple voting keys).
+        // ```cddl
+        //      [+delegation]
+        //      delegation = [$cip36_vote_pub_key, $weight]
+        //      $cip36_vote_pub_key /= bytes .size 32
+        // ```
+        minicbor::data::Type::Array => {
+            is_cip36 = true;
+            let len =
+                decode_array_len(d, "CIP36 Key Registration voting key, multiple voting keys")?;
+            for _ in 0..len {
+                let len = decode_array_len(d, "CIP36 Key Registration voting key, delegations")?;
+                // This fixed array should be a length of 2 (voting key, weight).
+                if len != 2 {
+                    return Err(decode::Error::message(format!(
+                        "Invalid length for CIP36 Key Registration voting key delegations, expected 2, got {len}"
+                    )));
+                }
+                // The first entry.
+                let pub_key = decode_bytes(
+                    d,
+                    "CIP36 Key Registration voting key, delegation array first entry (voting public key)",
+                )?;
+                // The second entry.
+                let weight: u32 = decode_helper(
+                    d,
+                    "CIP36 Key Registration voting key, delegation array second entry (weight)",
+                    &mut (),
+                )?;
+
+                let vk = voting_pk_vec_to_verifying_key(&pub_key).map_err(|e| {
+                    decode::Error::message(format!(
+                        "CIP36 Key Registration voting key, multiple voting keys, {e}"
+                    ))
+                })?;
+
+                voting_keys.push(VotingPubKey {
+                    voting_pk: vk,
+                    weight,
+                });
+            }
+        },
+        _ => {
+            return Err(decode::Error::message(
+                "Invalid datatype for CIP36 Key Registration voting key",
+            ))
+        },
+    }
+    Ok((is_cip36, voting_keys))
+}
+
+/// Helper function for converting `&[u8]` to `VerifyingKey`.
+fn voting_pk_vec_to_verifying_key(pub_key: &[u8]) -> anyhow::Result<VerifyingKey> {
+    VerifyingKey::from_bytes(
+        pub_key
+            .try_into()
+            .map_err(|_| anyhow::anyhow!("Invalid verifying key length"))?,
+    )
+    .map_err(|_| anyhow::anyhow!("Failed to convert to VerifyingKey"))
```

Comment:
Why do we want to discard the original error? It is very easy to preserve it with `anyhow` by using the `Context` crate:
```rust
pub_key.try_into().context("Invalid verifying key length")?
```
This way we can have both a higher level error description and the original cause of an error.